File: androidApp/src/main/kotlin/org/coreypett/fullstack/features/marketdetails/components/RecentTradesSection.kt
package org.coreypett.fullstack.features.marketdetails.components

import android.annotation.SuppressLint
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.animation.expandVertically
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.shrinkVertically
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.draw.clip
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.viewinterop.AndroidView
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.coreypett.fullstack.market.MR
import org.coreypett.fullstack.market.model.MarketSymbol
import org.coreypett.fullstack.market.trade.model.RecentTradesSide
import org.coreypett.fullstack.market.trade.presentation.RecentTradesRowDisplay
import org.coreypett.fullstack.market.trade.presentation.RecentTradesStatus
import org.coreypett.fullstack.market.trade.presentation.RecentTradesViewState
import org.coreypett.fullstack.support.toComposeColor
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Composable
fun RecentTradesSection(
    market: MarketSymbol,
    viewState: RecentTradesViewState,
) {
    if (!viewState.hasTrades) {
        if (viewState.status == RecentTradesStatus.Failed) {
            EmptyState(message = viewState.centerMessage ?: "Loading trades")
        } else {
            RecentTradesSkeleton(market = market)
        }
        return
    }

    var selectedExplorerUrl by remember { mutableStateOf<String?>(null) }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(0.dp),
    ) {
        RecentTradesHeader(market = market)
        viewState.recentTrades.take(18).forEach { recentTrade ->
            key(recentTrade.animationKey()) {
                RecentTradesAnimatedRow(
                    recentTrade = recentTrade,
                    onOpenTransaction = {
                        selectedExplorerUrl = recentTrade.hyperliquidExplorerUrl()
                    },
                )
            }
        }
    }

    selectedExplorerUrl?.let { url ->
        HyperliquidExplorerDialog(
            url = url,
            onDismiss = { selectedExplorerUrl = null },
        )
    }
}

@Composable
private fun RecentTradesSkeleton(
    market: MarketSymbol,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(0.dp),
    ) {
        RecentTradesHeader(market = market)
        repeat(7) { seed ->
            SkeletonRecentTradeRow(seed = seed)
        }
    }
}

@Composable
private fun SkeletonRecentTradeRow(
    seed: Int,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(30.dp)
            .background(MR.colors.app_background.toComposeColor())
            .padding(horizontal = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        SkeletonCell(
            width = recentTradePriceSkeletonWidth(seed),
            alignment = Alignment.CenterStart,
        )
        Box(
            modifier = Modifier.weight(1f),
            contentAlignment = Alignment.Center,
        ) {
            Box(
                modifier = Modifier.width(RecentTradesSizeColumnWidth),
                contentAlignment = Alignment.CenterStart,
            ) {
                SkeletonBlock(
                    width = recentTradeSizeSkeletonWidth(seed),
                    height = 13.dp,
                    cornerRadius = 3.dp,
                )
            }
        }
        SkeletonCell(
            width = 58.dp,
            alignment = Alignment.CenterEnd,
        )
    }
}

@Composable
private fun RowScope.SkeletonCell(
    width: Dp,
    alignment: Alignment,
) {
    Box(
        modifier = Modifier.weight(1f),
        contentAlignment = alignment,
    ) {
        SkeletonBlock(
            width = width,
            height = 13.dp,
            cornerRadius = 3.dp,
        )
    }
}

@Composable
private fun RecentTradesHeader(
    market: MarketSymbol,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 4.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        HeaderCell("Price(USD)", TextAlign.Start)
        SizeHeaderCell("Size (${market.displayName})")
        HeaderCell("Time", TextAlign.End)
    }
}

@Composable
private fun RowScope.HeaderCell(
    text: String,
    textAlign: TextAlign,
) {
    Text(
        text = text,
        modifier = Modifier.weight(1f),
        color = MR.colors.text_tertiary.toComposeColor(),
        fontSize = 11.sp,
        textAlign = textAlign,
    )
}

@Composable
private fun RowScope.SizeHeaderCell(
    text: String,
) {
    Box(
        modifier = Modifier.weight(1f),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = text,
            modifier = Modifier.width(RecentTradesSizeColumnWidth),
            color = MR.colors.text_tertiary.toComposeColor(),
            fontSize = 11.sp,
            textAlign = TextAlign.Start,
            maxLines = 1,
        )
    }
}

@Composable
private fun RecentTradesAnimatedRow(
    recentTrade: RecentTradesRowDisplay,
    onOpenTransaction: () -> Unit,
) {
    val animationKey = recentTrade.animationKey()
    var isVisible by remember(animationKey) { mutableStateOf(false) }
    val flashAlpha = remember(animationKey) { Animatable(RecentTradeFlashAlpha) }

    LaunchedEffect(animationKey) {
        isVisible = true
        flashAlpha.snapTo(RecentTradeFlashAlpha)
        flashAlpha.animateTo(
            targetValue = 0f,
            animationSpec = tween(durationMillis = 850),
        )
    }

    AnimatedVisibility(
        visible = isVisible,
        enter = slideInVertically(
            animationSpec = tween(durationMillis = 260),
            initialOffsetY = { -it / 2 },
        ) + expandVertically(
            animationSpec = tween(durationMillis = 260),
        ) + fadeIn(
            animationSpec = tween(durationMillis = 180),
        ),
        exit = slideOutVertically(
            animationSpec = tween(durationMillis = 180),
            targetOffsetY = { it / 3 },
        ) + shrinkVertically(
            animationSpec = tween(durationMillis = 180),
        ) + fadeOut(
            animationSpec = tween(durationMillis = 120),
        ),
    ) {
        RecentTradesRow(
            recentTrade = recentTrade,
            flashAlpha = flashAlpha.value,
            onOpenTransaction = onOpenTransaction,
        )
    }
}

@Composable
private fun RecentTradesRow(
    recentTrade: RecentTradesRowDisplay,
    flashAlpha: Float,
    onOpenTransaction: () -> Unit,
) {
    val sideColor = when (recentTrade.side) {
        RecentTradesSide.Buy -> MR.colors.bid.toComposeColor()
        RecentTradesSide.Sell -> MR.colors.ask.toComposeColor()
        RecentTradesSide.Unknown -> MR.colors.text_secondary.toComposeColor()
    }
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(30.dp)
            .background(MR.colors.app_background.toComposeColor())
            .background(sideColor.copy(alpha = flashAlpha))
            .padding(horizontal = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Text(
            text = recentTrade.priceText,
            modifier = Modifier.weight(1f),
            color = sideColor,
            fontSize = 13.sp,
            fontFamily = FontFamily.Monospace,
            textAlign = TextAlign.Start,
            maxLines = 1,
        )
        Box(
            modifier = Modifier.weight(1f),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = recentTrade.sizeText,
                modifier = Modifier.width(RecentTradesSizeColumnWidth),
                color = MR.colors.text_primary.toComposeColor(),
                fontSize = 13.sp,
                fontFamily = FontFamily.Monospace,
                textAlign = TextAlign.Start,
                maxLines = 1,
            )
        }
        Row(
            modifier = Modifier.weight(1f),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp, Alignment.End),
        ) {
            Text(
                text = recentTrade.timeMillis.toTradeTime(),
                color = MR.colors.text_secondary.toComposeColor(),
                fontSize = 13.sp,
                fontFamily = FontFamily.Monospace,
                textAlign = TextAlign.End,
                maxLines = 1,
            )
            TransactionExplorerButton(onClick = onOpenTransaction)
        }
    }
}

@Composable
private fun TransactionExplorerButton(
    onClick: () -> Unit,
) {
    val iconColor = MR.colors.text_secondary.toComposeColor()
    Box(
        modifier = Modifier
            .size(22.dp)
            .clip(CircleShape)
            .background(MR.colors.app_panel.toComposeColor().copy(alpha = 0.82f))
            .clickable(onClick = onClick)
            .semantics { contentDescription = "Open transaction in Hyperliquid explorer" },
        contentAlignment = Alignment.Center,
    ) {
        Canvas(modifier = Modifier.size(12.dp)) {
            val strokeWidth = 1.5.dp.toPx()
            val inset = 2.dp.toPx()
            val start = androidx.compose.ui.geometry.Offset(inset, size.height - inset)
            val end = androidx.compose.ui.geometry.Offset(size.width - inset, inset)

            drawLine(
                color = iconColor,
                start = start,
                end = end,
                strokeWidth = strokeWidth,
                cap = StrokeCap.Round,
            )
            drawLine(
                color = iconColor,
                start = androidx.compose.ui.geometry.Offset(size.width - inset, inset),
                end = androidx.compose.ui.geometry.Offset(size.width - inset, size.height * 0.42f),
                strokeWidth = strokeWidth,
                cap = StrokeCap.Round,
            )
            drawLine(
                color = iconColor,
                start = androidx.compose.ui.geometry.Offset(size.width - inset, inset),
                end = androidx.compose.ui.geometry.Offset(size.width * 0.58f, inset),
                strokeWidth = strokeWidth,
                cap = StrokeCap.Round,
            )
        }
    }
}

@SuppressLint("SetJavaScriptEnabled")
@Composable
private fun HyperliquidExplorerDialog(
    url: String,
    onDismiss: () -> Unit,
) {
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth(0.94f)
                .fillMaxHeight(0.86f)
                .clip(RoundedCornerShape(12.dp))
                .background(MR.colors.app_background.toComposeColor()),
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(44.dp)
                    .background(MR.colors.app_panel.toComposeColor())
                    .padding(start = 14.dp, end = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                val closeColor = MR.colors.text_secondary.toComposeColor()
                Text(
                    text = "Hyperliquid",
                    modifier = Modifier.weight(1f),
                    color = MR.colors.text_primary.toComposeColor(),
                    fontSize = 14.sp,
                    maxLines = 1,
                )
                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .clickable(onClick = onDismiss)
                        .semantics { contentDescription = "Close explorer" },
                    contentAlignment = Alignment.Center,
                ) {
                    Canvas(modifier = Modifier.size(12.dp)) {
                        val strokeWidth = 1.5.dp.toPx()
                        drawLine(
                            color = closeColor,
                            start = androidx.compose.ui.geometry.Offset(0f, 0f),
                            end = androidx.compose.ui.geometry.Offset(size.width, size.height),
                            strokeWidth = strokeWidth,
                            cap = StrokeCap.Round,
                        )
                        drawLine(
                            color = closeColor,
                            start = androidx.compose.ui.geometry.Offset(size.width, 0f),
                            end = androidx.compose.ui.geometry.Offset(0f, size.height),
                            strokeWidth = strokeWidth,
                            cap = StrokeCap.Round,
                        )
                    }
                }
            }

            AndroidView(
                modifier = Modifier
                    .fillMaxWidth()
                    .weight(1f),
                factory = { context ->
                    WebView(context).apply {
                        webViewClient = WebViewClient()
                        settings.javaScriptEnabled = true
                        settings.domStorageEnabled = true
                        loadUrl(url)
                    }
                },
                update = { webView ->
                    if (webView.url != url) {
                        webView.loadUrl(url)
                    }
                },
            )
        }
    }
}

private fun Long.toTradeTime(): String {
    return SimpleDateFormat("HH:mm:ss", Locale.US).format(Date(this))
}

private fun RecentTradesRowDisplay.hyperliquidExplorerUrl(): String =
    "https://app.hyperliquid.xyz/explorer/tx/$transactionHash"

private fun RecentTradesRowDisplay.animationKey(): String =
    "$tradeId-$transactionHash"

private val RecentTradesSizeColumnWidth = 64.dp
private const val RecentTradeFlashAlpha = 0.22f

private val RecentTradesPriceSkeletonWidths = listOf(72.dp, 84.dp, 64.dp, 78.dp)
private val RecentTradesSizeSkeletonWidths = listOf(34.dp, 46.dp, 38.dp, 52.dp)

private fun recentTradePriceSkeletonWidth(seed: Int): Dp =
    RecentTradesPriceSkeletonWidths[seed % RecentTradesPriceSkeletonWidths.size]

private fun recentTradeSizeSkeletonWidth(seed: Int): Dp =
    RecentTradesSizeSkeletonWidths[seed % RecentTradesSizeSkeletonWidths.size]

File: iosApp/iosApp/Features/MarketDetails/Components/RecentTradesSection.swift
import Pow
@preconcurrency import SharedLogic
import SwiftUI

struct RecentTradesState {
    let status: RecentTradesStatus
    let statusLabel: String
    let centerMessage: String?
    let rows: [RecentTradesRow]
    let isLoading: Bool

    var hasRows: Bool {
        !rows.isEmpty
    }

    static let preview = RecentTradesState(
        status: .live,
        statusLabel: "Live",
        centerMessage: nil,
        rows: [
            RecentTradesRow(side: .buy, sideText: "Buy", priceText: "$69,125.50", sizeText: "0.42", transactionHash: "0xabc123def4567890", timeMillis: 1_710_000_000_000, tradeId: 42),
            RecentTradesRow(side: .sell, sideText: "Sell", priceText: "$69,124", sizeText: "0.25", transactionHash: "0xdef456abc1237890", timeMillis: 1_710_000_000_100, tradeId: 43),
            RecentTradesRow(side: .buy, sideText: "Buy", priceText: "$69,126", sizeText: "0.18", transactionHash: "0x7890abc123def456", timeMillis: 1_710_000_000_200, tradeId: 44),
        ],
        isLoading: false
    )

    static let loading = RecentTradesState(
        status: .connecting,
        statusLabel: "Connecting",
        centerMessage: nil,
        rows: [],
        isLoading: true
    )

    init(shared: RecentTradesViewState) {
        self.status = shared.status
        self.statusLabel = shared.statusLabel
        self.centerMessage = shared.centerMessage
        self.rows = shared.recentTrades.map(RecentTradesRow.init)
        self.isLoading = shared.status == .connecting && !shared.hasTrades
    }

    init(
        status: RecentTradesStatus,
        statusLabel: String,
        centerMessage: String?,
        rows: [RecentTradesRow],
        isLoading: Bool
    ) {
        self.status = status
        self.statusLabel = statusLabel
        self.centerMessage = centerMessage
        self.rows = rows
        self.isLoading = isLoading
    }
}

struct RecentTradesRow: Hashable, Identifiable {
    let side: RecentTradesSide
    let sideText: String
    let priceText: String
    let sizeText: String
    let transactionHash: String
    let timeMillis: Int64
    let tradeId: Int64

    var id: String {
        "\(tradeId)-\(transactionHash)"
    }

    var timeText: String {
        Self.timeFormatter.string(from: Date(timeIntervalSince1970: Double(timeMillis) / 1_000))
    }

    init(shared: RecentTradesRowDisplay) {
        self.side = shared.side
        self.sideText = shared.sideText
        self.priceText = shared.priceText
        self.sizeText = shared.sizeText
        self.transactionHash = shared.transactionHash
        self.timeMillis = shared.timeMillis
        self.tradeId = shared.tradeId
    }

    init(
        side: RecentTradesSide,
        sideText: String,
        priceText: String,
        sizeText: String,
        transactionHash: String,
        timeMillis: Int64,
        tradeId: Int64
    ) {
        self.side = side
        self.sideText = sideText
        self.priceText = priceText
        self.sizeText = sizeText
        self.transactionHash = transactionHash
        self.timeMillis = timeMillis
        self.tradeId = tradeId
    }

    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss"
        return formatter
    }()
}

struct RecentTradesSection: View {
    let market: MarketSymbol
    let recentTrades: RecentTradesState

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if recentTrades.isLoading {
                RecentTradesSkeleton(market: market)
                    .transition(.movingParts.blur.combined(with: .opacity))
            } else if recentTrades.hasRows {
                RecentTradesList(market: market, rows: recentTrades.rows)
                    .transition(.movingParts.blur.combined(with: .opacity))
            } else {
                CenterMessage(text: recentTrades.centerMessage ?? "Connecting")
                    .frame(height: 180)
                    .transition(.movingParts.blur.combined(with: .opacity))
            }
        }
        .animation(.easeInOut(duration: 0.22), value: recentTrades.isLoading)
    }
}

private struct RecentTradesSkeleton: View {
    let market: MarketSymbol

    var body: some View {
        VStack(spacing: 0) {
            RecentTradesHeaderRow(market: market)

            ForEach(0..<7, id: \.self) { index in
                RecentTradesSkeletonRow(seed: index)
            }
        }
    }
}

private struct RecentTradesSkeletonRow: View {
    let seed: Int

    var body: some View {
        HStack(spacing: 10) {
            SkeletonBlock(width: priceWidth(for: seed), height: 13, cornerRadius: 3)
                .frame(maxWidth: .infinity, alignment: .leading)

            SkeletonBlock(width: sizeWidth(for: seed), height: 13, cornerRadius: 3)
                .frame(maxWidth: .infinity, alignment: .center)

            SkeletonBlock(width: 58, height: 13, cornerRadius: 3)
                .frame(maxWidth: .infinity, alignment: .trailing)
        }
        .frame(height: 30)
        .padding(.horizontal, 4)
    }

    private func priceWidth(for seed: Int) -> CGFloat {
        [72, 84, 64, 78][seed % 4]
    }

    private func sizeWidth(for seed: Int) -> CGFloat {
        [34, 46, 38, 52][seed % 4]
    }
}

private struct RecentTradesList: View {
    let market: MarketSymbol
    let rows: [RecentTradesRow]
    @Environment(\.openURL) private var openURL

    var body: some View {
        LazyVStack(spacing: 0) {
            RecentTradesHeaderRow(market: market)

            ForEach(Array(rows.prefix(18).enumerated()), id: \.element) { _, row in
                RecentTradesListRow(row: row) {
                    openURL(row.hyperliquidExplorerURL)
                }
            }
        }
    }
}

private struct RecentTradesHeaderRow: View {
    let market: MarketSymbol

    var body: some View {
        HStack(spacing: 10) {
            RecentTradesHeaderCell(text: "Price(USD)", alignment: .leading)
            Text("Size (\(market.displayName))")
                .font(.system(size: 11))
                .foregroundStyle(MR.colors.shared.text_tertiary.swiftUIColor)
                .frame(width: recentTradesSizeColumnWidth, alignment: .leading)
                .frame(maxWidth: .infinity, alignment: .center)
                .lineLimit(1)
                .minimumScaleFactor(0.75)
            RecentTradesHeaderCell(text: "Time", alignment: .trailing)
        }
        .padding(.horizontal, 4)
        .padding(.vertical, 6)
    }
}

private struct RecentTradesHeaderCell: View {
    let text: String
    let alignment: Alignment
    var width: CGFloat?

    var body: some View {
        Text(text)
            .font(.system(size: 11))
            .foregroundStyle(MR.colors.shared.text_tertiary.swiftUIColor)
            .frame(maxWidth: width == nil ? .infinity : nil, alignment: alignment)
            .frame(width: width, alignment: alignment)
            .lineLimit(1)
            .minimumScaleFactor(0.75)
    }
}

private struct RecentTradesListRow: View {
    let row: RecentTradesRow
    let onSelectHash: () -> Void

    var body: some View {
        HStack(spacing: 10) {
            Text(row.priceText)
                .foregroundStyle(recentTradesColor(for: row.side))
                .frame(maxWidth: .infinity, alignment: .leading)

            Text(row.sizeText)
                .foregroundStyle(MR.colors.shared.text_primary.swiftUIColor)
                .frame(width: recentTradesSizeColumnWidth, alignment: .leading)
                .frame(maxWidth: .infinity, alignment: .center)

            HStack(spacing: 6) {
                Text(row.timeText)
                    .foregroundStyle(MR.colors.shared.text_secondary.swiftUIColor)

                TransactionHashButton(action: onSelectHash)
            }
            .frame(maxWidth: .infinity, alignment: .trailing)
        }
        .font(.system(size: 13, design: .monospaced))
        .lineLimit(1)
        .minimumScaleFactor(0.72)
        .frame(height: 30)
        .padding(.horizontal, 4)
    }
}

private struct TransactionHashButton: View {
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: "arrow.up.right")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(MR.colors.shared.text_secondary.swiftUIColor)
                .frame(width: 22, height: 22)
                .background(MR.colors.shared.app_panel.swiftUIColor.opacity(0.82))
                .clipShape(Circle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Open transaction in Hyperliquid explorer")
    }
}

private func recentTradesColor(for side: RecentTradesSide) -> Color {
    switch side {
    case .buy:
        return MR.colors.shared.bid.swiftUIColor
    case .sell:
        return MR.colors.shared.ask.swiftUIColor
    case .unknown:
        return MR.colors.shared.text_secondary.swiftUIColor
    }
}

private extension RecentTradesRow {
    var hyperliquidExplorerURL: URL {
        URL(string: "https://app.hyperliquid.xyz/explorer/tx/\(transactionHash)")!
    }
}

private let recentTradesSizeColumnWidth: CGFloat = 64
